src/python/pants/backend/python/macros/poetry_requirements.py

```diff
@@ -55,7 +55,7 @@ def get_max_tilde(parsed_version: Version) -> str:
     return f"{major}.{minor}.0"
 
 
-def parse_str_version(proj_name: str, attributes: str, fp: str) -> str:
+def parse_str_version(proj_name: str, attributes: Any, fp: str, extras_string: str) -> str:
```

Comment:
For this function, it'll be string -- this is where the use of `isinstance` you mentioned this morning could come in, though it's also addressable with `TypedDict`. Mypy complains if the initial dictionary has a Union[str, list[str]] as the value type because you can't do certain string operations on a list (which, again, would only come about with the `extras` lookup but Mypy doesn't know that yet).